flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/GenericExactlyOnceSink.java

```diff
@@ -0,0 +1,173 @@
+import org.apache.flink.api.java.typeutils.TypeExtractor;
+import org.apache.flink.core.memory.DataInputView;
+import org.apache.flink.runtime.state.AbstractStateBackend;
+import org.apache.flink.runtime.state.StateHandle;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.api.watermark.Watermark;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.streaming.runtime.tasks.StreamTaskState;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Generic Sink that emits its input elements into an arbitrary backend. This sink is integrated with the checkpointing
+ * mechanism to provide exactly once semantics.
+ *
+ * Incoming records are stored within a {@link org.apache.flink.runtime.state.AbstractStateBackend}, and only committed if a
+ * checkpoint is completed. Should a job fail while the data is being committed, no exactly once guarantee can be made.
+ * @param <IN>
+ */
+public abstract class GenericExactlyOnceSink<IN extends Tuple> extends AbstractStreamOperator<IN> implements OneInputStreamOperator<IN, IN> {
+	private AbstractStateBackend.CheckpointStateOutputView out;
+	private TypeSerializer<IN> serializer;
+	protected TypeInformation<IN> typeInfo;
+
+	private ExactlyOnceState state = new ExactlyOnceState();
+
+	/**
+	 * Saves a handle in the state.
+	 * @param checkpointId
+	 * @throws IOException
+	 */
+	private void saveHandleInState(final long checkpointId) throws IOException {
+		//only add handle if a new OperatorState was created since the last snapshot/notify
+		if (out != null) {
+			out.writeByte(0); //EOF-byte
+			StateHandle<DataInputView> handle = out.closeAndGetHandle();
+			if (state.pendingHandles.containsKey(checkpointId)) {
+				state.pendingHandles.get(checkpointId).add(handle);
+			} else {
+				ArrayList<StateHandle<DataInputView>> list = new ArrayList<>();
+				list.add(handle);
+				state.pendingHandles.put(checkpointId, list);
+			}
+			out = null;
+		}
+	}
+
+	@Override
+	public StreamTaskState snapshotOperatorState(final long checkpointId, final long timestamp) throws Exception {
+		StreamTaskState taskState = super.snapshotOperatorState(checkpointId, timestamp);
+		saveHandleInState(checkpointId);
+		taskState.setFunctionState(state);
+		return taskState;
+	}
+
+	@Override
+	public void restoreState(StreamTaskState state, long recoveryTimestamp) throws Exception {
+		super.restoreState(state, recoveryTimestamp);
+		this.state = (ExactlyOnceState) state.getFunctionState();
+		out = null;
+	}
+
+	@Override
+	public void notifyOfCompletedCheckpoint(long checkpointId) throws Exception {
+		super.notifyOfCompletedCheckpoint(checkpointId);
+		saveHandleInState(checkpointId);
```

Comment:
I'm not so sure whether this is correct to do here. I think calling `saveHandleInState` will also include new elements which have arrived after the `checkpointId` barrier in the current `checkpointId` state. This would entail duplicate elements in case of a failure.